Unreal/Plugins/AirSim/Source/RenderRequest.cpp

```diff
@@ -49,38 +48,106 @@ void RenderRequest::RenderThreadScreenshotTask(RenderRequest::RenderResult &resu
+                */
+                result.pixels = buffer_pool_->GetBufferExactSize(
+                    width * height * 3);
+                // The stride should exactly fit the image.
+                result.stride = result.width * 3;
+
+                // Convert the data to BGR8
+                uint8* dst = result.pixels->data();
+                bgra8ToBgr8(dst, src, width, height, stride);
+
+                break;
+            }
 
         case PF_FloatRGBA:
-            result.pixels_as_float = true;
-            result.pixels_float = buffer_pool_float_->GetBufferExactSize(size);
-            FMemory::BigBlockMemcpy(result.pixels_float->data(), src, size);
-            break;
+            {
+                result.pixels_as_float = true;
+
+                /*
+                * Calculate the exact image size from the width, height, number
+                * of components, and size of each component
+                */
+                result.pixels_float = buffer_pool_float_->GetBufferExactSize(
+                    width * height * sizeof(float));
+                // The stride should exactly fit the image.
+                result.stride = result.width * sizeof(float);
+
+                /*
+                * We're using a depth buffer, where the depth is stored in the R
+                * component. FloatRGBA is misleading, as it should actually be
+                * Float16RGBA, or HalfFloatRGBA, because each component is
+                * represented by a 16 bit float and not a 32 bit float.
+                */
+                float* dst = result.pixels_float->data();
+                rgba16fToDepthFloat(dst, src, width, height, stride);
+
+                break;
+            }
 
         default:
             UE_LOG(LogTemp, Warning, TEXT("Unexpected pixel format: %d"), pixelFormat);
             break;
         }
     }
-    
+
     RHIUnlockTexture2D(fast_cap_texture, 0, false);
 
     std::unique_lock<std::mutex> lck(mtx_);
     fast_cap_done_ = true;
     cv_.notify_all();
 }
+
+void RenderRequest::bgra8ToBgr8(
+    uint8* dst,
+    const uint8* src,
+    uint32 width,
+    uint32 height,
+    uint32 stride)
+{
+    for (uint32 i = 0; i < width; i++)
+    {
+        for (uint32 j = 0; j < height; j++)
+        {
+            memcpy(
+                &dst[(i + j * width) * 3],
+                &src[i * 4 + j * stride],
+                3 * sizeof(uint8));
+        }
+    }
+}
+
+void RenderRequest::rgba16fToDepthFloat(
+    float* dst,
+    const uint8* src,
+    uint32 width,
+    uint32 height,
+    uint32 stride)
+{
+    for (uint32 i = 0; i < width; i++)
+    {
+        for (uint32 j = 0; j < height; j++)
+        {
+            // Decode float rgba
+            int index0 = i * 8 + j * stride;
+            FFloat16 val;
+            val.Encoded = src[index0 + 0] | (src[index0 + 1] << 8);
+
+            /* Convert the value to meters and return it. */
+            dst[i + j * width] = val / 100.0f;
```

Comment:
Should it be `val.GetFloat()` here?

Reply:
I don't think that is necessary since `operator float()` is defined for FFloat16 so val should be implicitly converted, but I can change it if you think it is more readable that way.

Reply:
Nah, seems okay to me, fairly obvious what's happening here